enterprise/server/hit_tracker_client/hit_tracker_client.go

```diff
@@ -2,57 +2,280 @@ package hit_tracker_client
+	start           time.Time
+	client          hitpb.HitTrackerServiceClient
 }
 
 func (t *TransferTimer) CloseWithBytesTransferred(bytesTransferredCache, bytesTransferredClient int64, compressor repb.Compressor_Value, serverLabel string) error {
+	hit := &hitpb.CacheHit{
+		RequestMetadata: t.requestMetadata,
+		Resource: &rspb.ResourceName{
+			Digest:    t.digest,
+			CacheType: rspb.CacheType_CAS,
+		},
+		SizeBytes: bytesTransferredClient,
+		Duration:  durationpb.New(time.Since(t.start)),
+	}
+	t.enqueueFn(t.ctx, t.requestMetadata, hit)
 	return nil
 }
 
 func (t *TransferTimer) Record(bytesTransferred int64, duration time.Duration, compressor repb.Compressor_Value) error {
-	return nil
+	return status.InternalError("Unxpected call to hit_tracker_client.Record()")
+}
+
+func (h *HitTrackerClient) TrackDownload(digest *repb.Digest) interfaces.TransferTimer {
+	return &TransferTimer{
+		ctx:             h.ctx,
+		enqueueFn:       h.enqueueFn,
+		requestMetadata: h.requestMetadata,
+		digest:          digest,
+		start:           time.Now(),
+		client:          h.client,
+	}
+}
+
+// Writes hit the backing cache, so no need to report on hit-tracking.
+func (h *HitTrackerClient) TrackUpload(digest *repb.Digest) interfaces.TransferTimer {
+	return &NoOpTransferTimer{}
+}
+
+func (h *HitTrackerFactory) start() {
+	for {
+		select {
+		case <-h.quit:
+			return
+		case <-h.ticker:
+			h.sendTrackRequests(context.Background())
+		}
+	}
 }
 
-func (h *HitTracker) TrackDownload(d *repb.Digest) interfaces.TransferTimer {
-	return &TransferTimer{}
+func (h *HitTrackerFactory) sendTrackRequests(ctx context.Context) int {
+	i := 0
+	groups := make([]groupID, len(h.hitsByGroup))
+	h.mu.Lock()
+	for group := range h.hitsByGroup {
+		groups[i] = group
+		i++
+	}
+	h.mu.Unlock()
+
+	for _, groupID := range groups {
+		h.mu.Lock()
+		headers := h.hitsByGroup[groupID].authHeaders
+		trackRequest := hitpb.TrackRequest{Hits: h.hitsByGroup[groupID].hits}
+		h.hitsByGroup[groupID] = &cacheHits{}
+		h.mu.Unlock()
+
+		if len(trackRequest.Hits) > 0 {
+			ctx = authutil.AddAuthHeadersToContext(ctx, headers, h.authenticator)
+			_, err := h.client.Track(ctx, &trackRequest)
+			metrics.RemoteAtimeUpdatesSent.With(
+				prometheus.Labels{
+					metrics.GroupID:     string(groupID),
+					metrics.StatusLabel: gstatus.Code(err).String(),
+				}).Inc()
+			if err != nil {
+				log.CtxWarningf(ctx, "Error sending Track request to record cache hit-tracking state group %s: %v", groupID, err)
+			}
+		}
+	}
+	return 0
 }
 
-func (h *HitTracker) TrackUpload(d *repb.Digest) interfaces.TransferTimer {
-	return &TransferTimer{}
+func (h *HitTrackerFactory) shutdown(ctx context.Context) error {
+	close(h.quit)
+
+	// Make a best-effort attempt to flush pending updates.
+	// TODO(iain): we could do something fancier here if necessary, like
+	// fire-and-forget these RPCs with a rate-limiter. Let's try this for now.
+	for h.sendTrackRequests(ctx) > 0 {
+	}
+
+	return nil
```

Comment:
I think this shutdown would ideally wait until the gRPC server is either fully gracefully shutdown (or force-shutdown), otherwise we can lose hits for RPCs that are in-flight during the 30s graceful shutdown window.

I'd also be OK w/ the current approach for now, but maybe there's some way we can log how many requests are still in the buffer just before the server exits, so we have some way of knowing how many hits we're losing?

Reply:
Ah, good catch! I changed it to just send the hit-tracking RPCs synchronously in this case because they're kind of important and this is an edge case where the server is still serving traffic after its lame-duck period has passed. Does that seem reasonable to you?

Reply:
sgtm - we use a similar approach in redisutil.CommandBuffer and it seems to work well